crates/circuit/src/symbol_expr.rs

```diff
@@ -0,0 +1,3298 @@
+                    },
+                },
+                BinaryOp::Sub => match lhs.neg_opt() {
+                    Some(ln) => Some(_add(ln, rhs.as_ref().clone())),
+                    None => Some(_add(_neg(lhs.as_ref().clone()), rhs.as_ref().clone())),
+                },
+                BinaryOp::Mul => match lhs.neg_opt() {
+                    Some(ln) => Some(_mul(ln, rhs.as_ref().clone())),
+                    None => rhs.neg_opt().map(|rn| _mul(lhs.as_ref().clone(), rn)),
+                },
+                BinaryOp::Div => match lhs.neg_opt() {
+                    Some(ln) => Some(_div(ln, rhs.as_ref().clone())),
+                    None => rhs.neg_opt().map(|rn| _div(lhs.as_ref().clone(), rn)),
+                },
+                _ => None,
+            },
+            _ => None,
+        }
+    }
+
+    /// optimize the equation
+    pub fn optimize(&self) -> SymbolExpr {
+        match self {
+            SymbolExpr::Value(_) => self.clone(),
+            SymbolExpr::Symbol(_) => self.clone(),
+            SymbolExpr::Unary { op, expr } => {
+                let opt = expr.optimize();
+                match op {
+                    UnaryOp::Neg => match opt.neg_opt() {
+                        Some(e) => e,
+                        None => _neg(opt),
+                    },
+                    _ => SymbolExpr::Unary {
+                        op: op.clone(),
+                        expr: Box::new(opt),
+                    },
+                }
+            }
+            SymbolExpr::Binary { op, lhs, rhs } => {
+                let opt_lhs = lhs.optimize();
+                let opt_rhs = rhs.optimize();
+                match op {
+                    BinaryOp::Add => match opt_lhs.add_opt(&opt_rhs, true) {
+                        Some(e) => e,
+                        None => _add(opt_lhs, opt_rhs),
+                    },
+                    BinaryOp::Sub => match opt_lhs.sub_opt(&opt_rhs, true) {
+                        Some(e) => e,
+                        None => _sub(opt_lhs, opt_rhs),
+                    },
+                    BinaryOp::Mul => match opt_lhs.mul_opt(&opt_rhs, true) {
+                        Some(e) => e,
+                        None => _mul(opt_lhs, opt_rhs),
+                    },
+                    BinaryOp::Div => match opt_lhs.div_opt(&opt_rhs, true) {
+                        Some(e) => e,
+                        None => _div(opt_lhs, opt_rhs),
+                    },
+                    BinaryOp::Pow => _pow(opt_lhs, opt_rhs),
+                }
+            }
+        }
+    }
+}
+
+impl Add for SymbolExpr {
+    type Output = SymbolExpr;
+    fn add(self, rhs: Self) -> SymbolExpr {
+        &self + &rhs
+    }
+}
+
+impl Add for &SymbolExpr {
+    type Output = SymbolExpr;
+    fn add(self, rhs: Self) -> SymbolExpr {
+        match self.add_opt(rhs, false) {
+            Some(e) => e,
+            None => _add(self.clone(), rhs.clone()),
+        }
+    }
+}
```

Comment:
I'm not keen on defining these math ops for the reference types, I think.

It feels like we're giving `SymbolExpr` the semantics of `Copy` without it actually being `Copy`. I think it's better to just have the impls for `SymbolExpr` and if the user doesn't want the target consumed, they have to `clone` first.